lightning/src/ln/channelmanager.rs

```diff
@@ -2639,6 +2724,24 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			}
 
 			claimed_any_htlcs
+		} else if let Some(htlc) = removed_keysend_source {
+			assert!(removed_source.is_none());
```

Comment:
I believe this is a DoS risk - an intermediate node could send us a payment after seeing a keysend payment go by (well, they have to guess its keysend, but they maybe can) and cause us to panic here. More generally, the fact that you have two maps but expect a given key to exist only in one of the two screams "make it one map with an enum value" to me.